vhost-device-template/src/vhu_foo.rs

```diff
@@ -0,0 +1,394 @@
+//
+// SPDX-License-Identifier: Apache-2.0 or BSD-3-Clause
+
+use log::warn;
+use std::{
+    convert,
+    io::{self, Result as IoResult},
+};
+
+use thiserror::Error as ThisError;
+use vhost::vhost_user::message::{VhostUserProtocolFeatures, VhostUserVirtioFeatures};
+use vhost_user_backend::{VhostUserBackendMut, VringRwLock, VringT};
+use virtio_bindings::bindings::virtio_config::{VIRTIO_F_NOTIFY_ON_EMPTY, VIRTIO_F_VERSION_1};
+use virtio_bindings::bindings::virtio_ring::{
+    VIRTIO_RING_F_EVENT_IDX, VIRTIO_RING_F_INDIRECT_DESC,
+};
+use virtio_queue::{DescriptorChain, QueueOwnedT};
+use vm_memory::{GuestAddressSpace, GuestMemoryAtomic, GuestMemoryLoadGuard, GuestMemoryMmap};
+use vmm_sys_util::epoll::EventSet;
+use vmm_sys_util::eventfd::{EventFd, EFD_NONBLOCK};
+
+use crate::FooInfo;
+
+const QUEUE_SIZE: usize = 1024;
+const NUM_QUEUES: usize = 1;
+
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, ThisError)]
+/// Errors related to vhost-device-foo daemon.
+pub(crate) enum Error {
+    #[error("Failed to handle event, didn't match EPOLLIN")]
+    HandleEventNotEpollIn,
+    #[error("Failed to handle unknown event")]
+    HandleEventUnknown,
+    #[error("Descriptor not found")]
+    DescriptorNotFound,
+    #[error("Failed to send notification")]
+    NotificationFailed,
+    #[error("Failed to create new EventFd")]
+    EventFdFailed,
+}
+
+impl convert::From<Error> for io::Error {
+    fn from(e: Error) -> Self {
+        io::Error::new(io::ErrorKind::Other, e)
+    }
+}
+
+pub(crate) struct VhostUserFooBackend {
+    info: FooInfo,
+    event_idx: bool,
+    pub exit_event: EventFd,
+    mem: Option<GuestMemoryLoadGuard<GuestMemoryMmap>>,
+}
+
+type FooDescriptorChain = DescriptorChain<GuestMemoryLoadGuard<GuestMemoryMmap<()>>>;
+
+impl VhostUserFooBackend {
+    pub fn new(info: FooInfo) -> Result<Self> {
+        Ok(VhostUserFooBackend {
+            info,
+            event_idx: false,
+            exit_event: EventFd::new(EFD_NONBLOCK).map_err(|_| Error::EventFdFailed)?,
+            mem: None,
+        })
+    }
+
+    /// Process the requests in the vring and dispatch replies
+    fn process_requests(
+        &mut self,
+        requests: Vec<FooDescriptorChain>,
+        _vring: &VringRwLock,
+    ) -> Result<()> {
+        if requests.is_empty() {
+            println!("No pending requests\n");
+            return Ok(());
+        }
+
+        // Iterate over each FOO request.
+        for desc_chain in requests.clone() {
+            let counter = self.info.counter();
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+
+            println!(
+                "Request number: {} contains {} descriptors\n",
+                counter,
+                descriptors.len(),
+            );
+
+            for (i, desc) in descriptors.iter().enumerate() {
+                // One can also read/write the contents of the descriptors here.
```

Comment:
Is there any way to know which descriptions are read and which are write? Is it all implied from the spec? Are write descriptors always after the read ones?

Reply:
> Is there any way to know which descriptions are read and which are write?

The DescriptorChain from vm-virtio has this iterator, but also a .readable() and .writable() iterators that only yield if the current cursor is readable/writable. But hopefully we can find a better API with the descriptor_utils merge...

> Are write descriptors always after the read ones?

yes. The spec mandates that.

Reply:
`descriptor.is_write_only()`

Reply:
So I think we could do with a slightly better comment at the head of the for() to bring this all together, maybe something like:

    We now can iterate of the set of descriptors and process the messages. There will be a number of read only descriptors containing messages as defined by the specification. If any replies are needed the driver should have placed one or more writable descriptors at the end for the device to use to reply.

Reply:
Done.